promql/quantile.go

```diff
@@ -357,22 +359,41 @@ func coalesceBuckets(buckets buckets) buckets {
 // bucket with the φ-quantile count, so breaking the monotonicity
 // guarantee causes bucketQuantile() to return undefined (nonsense) results.
 //
-// As a somewhat hacky solution, we calculate the "envelope" of the histogram
-// buckets, essentially removing any decreases in the count between successive
-// buckets. We return a bool to indicate if this monotonicity was forced or not.
-func ensureMonotonic(buckets buckets) bool {
-	forced := false
-	max := buckets[0].count
+// As a somewhat hacky solution, we first silently correct any numerically
+// insignificant (below the requested tolerance and likely to be from floating
+// point precision errors) differences between successive buckets regardless of
+// the direction. Then we calculate the "envelope" of the histogram buckets,
+// essentially removing any decreases in the count between successive buckets.
+// We return a bool to indicate if this monotonicity was forced or not, and
+// another bool to indicate if the precision errors were fixed or not.
+func ensureMonotonicAndCorrectPrecisionErrors(buckets buckets, tolerance float64) (bool, bool) {
+	var forcedMonotonic, fixedPrecision bool
+	prev := buckets[0].count
 	for i := 1; i < len(buckets); i++ {
-		switch {
-		case buckets[i].count > max:
-			max = buckets[i].count
-		case buckets[i].count < max:
-			buckets[i].count = max
-			forced = true
+		curr := buckets[i].count // Assumed always positive.
+		delta := math.Abs(curr - prev)
+		if delta == 0 {
+			// No correction needed if the counts are identical between buckets.
+			continue
+		}
+		if (delta / curr) <= tolerance {
```

Comment:
I'm afraid you have to take into account the edge cases dealt with by https://floating-point-gui.de/errors/comparison/ and subsequently https://github.com/beorn7/floats/blob/master/equal.go . Since histograms get scaled when we `rate` them, you might get very small values, so you might run into the problem with sub-normal values. You also need to treat the case where `curr` is zero.

Reply:
I modified https://github.com/prometheus/prometheus/pull/13153/files#diff-67fbbf5503bf54336628b956808a7fdb0a6de1e3593f3d69c145bc1c880a7b73R668 to look more like the code you linked, and made it take epsilon as an argument so the function could be reused. Now that we use that, we don't need to check for curr being zero because the function already considers that right?